### DISCUSSION TITLE
System blowing up - and running with polymers/peptides

### DISCUSSION THREAD

**Discussion by mayank-kohli** [luancarvalhomartins/PyAutoFEP#41] [Q&A]
Hi, I was following the tutorial that is provided while running the final bash script for the protein part I, am getting this error.

ERROR 1 [file FXR_12-FXR_74/protein/lambda0/min01/min01.mdp]:
  The largest distance between excluded atoms is 1.063 nm, which is larger
  than the cut-off distance. This will lead to missing long-range
  corrections in the forces and energies.

any comments on the same that why I am getting this error will be beneficial.

**Comment by luancarvalhomartins** [comment]
Did you run prepare_dual_topologies.py with the tutorial inputs? Your system is likely blowing up during the first minimization, which is odd. Would you mind checking that the system is sane (eg, by taking a look at it in pymol)?

**Reply by mayank-kohli** [reply]
I have used only the tutorial inputs, and I checked fullsystem.pdb in that I didn't found any error. 
I have some other questions as well, like is it necessary that I should setup up the slurm for the submission and is there any specific version compatibility with gromacs.

**Reply by mayank-kohli** [reply]
I was getting this error in the gromacs 2022 version, so, I tried running it in gromacs 2021 version I didn't got the same error but this time I got a lot of LINCS error and eventually the segment fault error.

**Reply by luancarvalhomartins** [reply]
Your system is blowing up. This is weird. Would you mind trying to run `prepare_dual_topology.py` again? Maybe the water/ion placement was problematic and there are clashing particles causing the blow up. I can't think of anything else causing this. If this won't fix the problem, would you mind sharing the offensive inputs so I can try studying it?

Also:
 - No, slurm isn't a requirement. You can also use Torque/PBS with `output_script_type=pbs` (SGE may work, untested) or no job scheduler at all with `output_script_type=bash` (so that MD will be runned using a simple bash script). Using `output_script_type=bash` will run everything in a single machine/node, though. 
 - Any MPI-enabled GROMACS version should work. I tested quite extensively with 2016, 2018, 2019, and 2020 series, but any version >= 5.XX _should_ work (ie, given what PyAutoFEP uses gmx for). A plumed-patched gmx_mpi supporting `-hrex` is required for HREX.  Note: your specific problem is not related to the GROMACS version.

**Reply by mayank-kohli** [reply]
I ran again from scratch with 2021 version of gromacs (MPI enabled) using the tutorial files it worked fine, but in the case of 2022 version of gromacs I don't know why I am still getting the same error again (haven't tried my own molecules yet) .
And 
I want to know if I have the larger molecules (e.g. peptides or other polymers), does any specific parameter is required for it?

**Reply by luancarvalhomartins** [reply]
I wouldn't expect that version 2022 to behave in a different way than 2021 for the same system. But I am glad to know that it worked.

I wrote PyAutoFEP mainly for small organic molecules,  but there is no hard limit that prevents it to work with larger molecules. I guess that the most relevant problem for peptides or nucleic acids is that you would need to use the macromolecule FF to describe the ligand, while PyAutoFEP expects a ligand topology as input. Here's an idea: use a toll like parmed to convert/export the peptide/DNA/RNA topology a single file. Then use it as lig topology input to `prepare_dual_topology.py`. It should work.

The MCS algorithm may take a long time for large molecules and you may want to use the 3d MCS (but the default, graph-based MCS may work as well). Passing the MCS as smiles can greatly decrease the `prepare_dual_topology.py`. 

PS: there is a long-term goal to add a decent support for peptides in PyAutoFEP. It will use some of the code that I am writing for the charged perturbations, so peptide-specific code is not seeing the light of the day any time soon. In case you wanna give the regular code a try using peptides, feel free to write me here or by email. I would love to help you (and test the code for this use case in the process).

**Reply by mayank-kohli** [reply]
Thank you so much for your advice.
At present I have a plan to run for long polymers like structures so I will be using the same protocol.
Later on I will be trying for the peptides.
I will surely be in touch with you.

**Reply by luancarvalhomartins** [reply]
Sure thing. It's luancarvalhomartins@gmail.com. Feel free to write me.

I guess polymers would actually be easier than peptides/DNA/RNA. On the other hand, supplying an MCS will very likely be required (see https://www.rdkit.org/docs/GettingStartedInPython.html#maximum-common-substructure).

**Reply by mayank-kohli** [reply]
Thank you for the suggestion I will look into MCS while preparing the systems.

**Comment by mayank-kohli** [comment]
Hello, I'm having a new problem where all of the simulations are running fine, but the 'rerun' step is not running for some reason.

**Reply by luancarvalhomartins** [reply]
What is in the logs? `gmx mdrun rerun` should run fine more often than the MD run itself. The only issue I can think about is #30. In case that's your issue, there is a workaround in the first post and the fixing for the underlying cause was reported upstream. In case that's your issue, feel free to comment at the #30 thread.

In case that's not your issue, please, attach or paste the relevant parts of your logs. 

_Note: rerun logs are stored in_`ligA-ligB/{protein,water}/md/rerun/rerun_structX_coordY.log`

**Reply by mayank-kohli** [reply]
There is no file in there, the rerun folder is empty.

**Reply by luancarvalhomartins** [reply]
If the run step completed fine and rerun didn't start, there should be some error in job the output (in case you are using slurm or PBS, it should have been redirected to a file). BTW, is `parallel` (https://www.gnu.org/software/parallel/) installed in the run node? The rerun step uses it.

The current run script should be more verbose and user-friendly. I am planning to rewrite it from ground up at some point.

**Reply by mayank-kohli** [reply]
parallel is not there, after installing its working now.

**Reply by mayank-kohli** [reply]
During the system preparation that is first step "pdb2gmx", Is there any specific reason you haven't included '-ignh' flag?

During the nvt step at lambda3 window I got the lincs error
[nvt_mdrun_log.log](https://github.com/luancarvalhomartins/PyAutoFEP/files/8630429/nvt_mdrun_log.log)

**Reply by luancarvalhomartins** [reply]
I didn't add `-ignh` to the `pdb2gmx` command line because I believe that the most common use case is to pass a protonated PDB file (eg, PROPKa or H++) to `prepare_dual_topology.py`. Using `-ignh` would prevent such use. You can preprocess you input PDB, saving to PDB, using `pdb2gmx`, if you really want `-ignh`.

Regarding the blown up, is hard to know exatcly what happened. Is 4698 a ligand atom? Maybe you want to double check the starting pose, including the dual-molecule (by opening `FullSystem.pdb` in eg, pymol or vmd). Maybe you need more lambda windows. Maybe you need a longer minimization or nve equilibration.

**Reply by mayank-kohli** [reply]
Thank you for the suggestions, I ll look into these  and come back to you.

**Reply by mayank-kohli** [reply]
I analyzed the files and the error is in the FullSystem.pdb, the dual molecule is not correct.
![image](https://user-images.githubusercontent.com/64058587/167103643-236ab3e1-303a-4915-8e78-355aae19d53f.png)

**Reply by luancarvalhomartins** [reply]
From this image it is not clear whether the dual-molecule is correct or not. Because in the dual-topology module, atoms from both A and B states are modelled, pymol will render all atoms, leading to incorrect structures. One way to assess the dual-topology is from the MCS data produced by `prepare_dual_topology.py`. Alternatively, I put together a script to extract ligands A and B from the ligand.itp and fullsystemXXX.pdb (I suggest you using it in the water leg). Then, you can inspect the `liga.pdb` and `ligb.pdb`.

[extract_dual_mol.tar.gz](https://github.com/luancarvalhomartins/PyAutoFEP/files/8645256/extract_dual_mol.tar.gz)

Anyway, from the structure you seem to be perturbing, you seem to be trying to perturb a quite large structure to smaller one. Perturbing this many atoms will likely require more than 12 lambda windows and maybe a longer equilibration. Also, make sure that the binding does allow this large perturbation.

**Reply by mayank-kohli** [reply]
So, I was trying with different combinations for my system in which got again a segment fault when I see the structure the atoms of the ligands is floating in the system, I have attached the image, I am not sure what is happening.
![2](https://user-images.githubusercontent.com/64058587/167987477-3548aee5-efd5-45d8-81e2-749729e8245b.PNG)
![1](https://user-images.githubusercontent.com/64058587/167987481-73bf2eed-59db-4b8a-b61c-0a1bd9dbabb6.PNG)

**Reply by luancarvalhomartins** [reply]
The disconnected CH2 maybe a artifact of the PyMOL representation (it draws bonds based on atomic distance) or maybe a problem in the dual-topology generated by PyAutoFEP. Is the second image a representation of the system after the crash or the input structure? If it's the latter, that's almost certainly a bug in `prepare_dual_topology.py`.

 Would you mind extracting the A and B molecules so we can better inspect what's going on? Or sharing the starting molecules and the full output of the `prepare_dual_topology.py` using `-vvvv` (debug mode). If your molecules are not public, feel free to reach me by email.

If you'd rather get a closer look yourself, look for the for the MCS info in `prepare_dual_topology.py` debug output (`-vvvv`). It will tell you how the dual-topology will be generated. Inspecting the ligand.itp will also hint for problems.
